Experimental/src/Ranges/Range.cs

```diff
@@ -0,0 +1,713 @@
+					value = "(" + value;
+				if (!IncludeEnd)
+					value = value + ")";
+
+				return value;
+			}
+
+			var startValue = string.Empty;
+			var endValue = string.Empty;
+
+			if (HasStart)
+            {
+                startValue = Start.ToString();
+                if (!IncludeStart)
+                    startValue = "(" + startValue + ")";
+            }
+
+            if (HasEnd)
+            {
+                endValue = End.ToString();
+                if (!IncludeEnd)
+                    endValue = "(" + endValue + ")";
+            }
+
+            return startValue + ".." + endValue;
+        }
+
+        private struct RangeValue
+        {
+            public readonly bool HasValue;
+            public readonly bool Included;
+            public readonly TValue Value;
+
+            public RangeValue(TValue value, bool hasValue, bool included)
+            {
+                Value = value;
+                HasValue = hasValue;
+                Included = included;
+            }
+
+            public override string ToString()
+            {
+                if (!HasValue)
+                    return string.Empty;
+                if (!Included)
+                    return string.Format("({0})", Value);
+                return Value.ToString();
+            }
+
+            public static RangeValue MinValue(RangeValue value1, RangeValue value2, bool isAnd)
+            {
+                if (!value1.HasValue)
+                    return value1;
+                if (!value2.HasValue)
+                    return value2;
+
+                var compare = value1.Value.CompareTo(value2.Value);
+                if (compare > 0)
+                    return value2;
+
+                var newIncluded = value1.Included;
+                if (compare == 0)
+                {
+                    if (isAnd)
+                        newIncluded = value2.Included && newIncluded;
+                    else
+                        newIncluded = value2.Included || newIncluded;
+                }
+
+                return new RangeValue(value1.Value, true, newIncluded);
+            }
+
+            public static RangeValue MaxValue(RangeValue value1, RangeValue value2, bool isAnd)
+            {
+                if (!value1.HasValue)
+                    return value1;
+                if (!value2.HasValue)
+                    return value2;
+
+                var compare = value1.Value.CompareTo(value2.Value);
+                if (compare < 0)
+                    return value2;
+
+                var newIncluded = value1.Included;
+                if (compare == 0)
+                {
+                    if (isAnd)
+                        newIncluded = value2.Included && newIncluded;
+                    else
+                        newIncluded = value2.Included || newIncluded;
+                }
+
+                return new RangeValue(value1.Value, true, newIncluded);
+            }
+        }
+    }
+
+	public static class RangeValueUtils
```

Comment:
There are no methods related to RangeValue, so the name RangeValueUtils looks inappropriate